pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,425 @@
+            case _:
+                ValueError(f'ConduitNode: __setitem__ type not supported')
+
+    def __str__(self):
+        self.lib.conduit_node_print(self)
+        return '\n'
+
+    def append(self):
+        ptr = self.lib.conduit_node_append(self)
+        return ConduitNode(self.lib, ptr, child=True)
+
+
+class AscentWrappers(LibWrapper):
+    _libname = 'ascent_mpi'
+
+    # Functions
+    _functions = [
+        (None, 'ascent_close', c_void_p),
+        (c_void_p, 'ascent_create', c_void_p),
+        (None, 'ascent_execute', c_void_p, c_void_p),
+        (None, 'ascent_open', c_void_p, c_void_p),
+        (None, 'ascent_publish', c_void_p, c_void_p),
+    ]
+
+
+class AscentPlugin(RegionMixin, BasePlugin):
+    name = 'ascent'
+    systems = ['*']
+    formulations = ['dual', 'std']
+
+    # Ascent filters that implicit require velocity expression
+    v_filter = ['gradient', 'qcriterion', 'vorticity']
+
+    def __init__(self, intg, cfgsect, suffix=None):
+        super().__init__(intg, cfgsect, suffix)
+
+        comm, rank, root = get_comm_rank_root()
+        self.nsteps = self.cfg.getint(cfgsect, 'nsteps')
+
+        # Underlying system
+        system = intg.system
+        self.dtype = intg.system.backend.fpdtype
+        self.ndims = intg.system.ndims
+
+        # Underlying system elements class
+        self.elementscls = system.elementscls
+
+        # Load conduit library
+        self.conduit = ConduitWrappers()
+
+        # Expressions to plot and configs
+        self._expr = {}
+        self._config_field(cfgsect)
+        self._config_scene(cfgsect)
+        self._config_pipeline(cfgsect)
+
+        # Gradient pre-processing
+        self._init_gradients(intg)
+
+        # Base directory for output and counter reset
+        self.basedir = self.cfg.getpath(cfgsect, 'basedir', '.', abs=True)
+        for sname, scfg in self._scene_cfg.items():
+            scfg['path_gen'] = file_path_gen(self.basedir, scfg['image_name'])
+            next(scfg['path_gen'])
+
+        # Set subdivision
+        self.divisor = self.cfg.getint(cfgsect, 'division', 
+                                       self.cfg.getint('solver', 'order'))
+        self.etypes_div = defaultdict(lambda: self.divisor)
+
+        # Share number of ele types per part
+        ndoms = len(system.ele_map)
+        ndoms = comm.gather(ndoms, root=root)
+        if rank == root:
+            ndoms = np.asarray(ndoms)
+        ndoms = comm.bcast(ndoms, root=root)
+        self.dom_0 = dom_0 = np.sum(ndoms[:rank])
+
+        # Generate conduit blueprint of mesh
+        mesh_n = ConduitNode(self.conduit)
+        self.ops = {}
+
+        for idx, ename, rgn in self._ele_regions:
+            if ename in ['pri', 'pyr']:
+                raise ValueError(f'Ascent: {ename} elements not supported')
+            eles = system.ele_map[ename]
+
+            dpf = f'domain_{dom_0 + idx}'
+
+            shape = subclass_where(BaseShape, name=ename)(eles.nspts, 
```

Comment:
Don't we have a helper method for this?

Reply:
Not that I'm aware of, this is the same techniques used to do the same thing in a number of places in pyfr

Reply:
`_get_shape`?

Reply:
Yep, forgot about that